libqtile/widget/openweather.py

```diff
@@ -0,0 +1,215 @@
+# -*- coding:utf-8 -*-
+# Copyright (c) 2020 Himanshu Chauhan
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import time
+from urllib.parse import urlencode
+
+from libqtile.widget import base
+from libqtile.widget.generic_poll_text import GenPollUrl
+
+# API access details at https://openweathermap.org/current
+QUERY_URL = "http://api.openweathermap.org/data/2.5/weather?"
+
+
+class OpenWeatherResponseError(Exception):
+    def __init__(self, resp_code, err_str=None):
+        self.resp_code = resp_code
+        self.err_str = err_str
+
+
+class OpenWeatherResponseParser(object):
+    def __init__(self, resp_json):
+        self.resp_json = resp_json
+        self.fjson = self.flatten_json(self.resp_json)
+        if int(self.fjson["cod"]) != 200:
+            raise OpenWeatherResponseError(int(self.fjson["cod"]))
+
+    def flatten_json(self, obj):
+        out = {}
+
+        def __inner(_json, name=''):
+            if type(_json) is dict:
+                for key, value in _json.items():
+                    __inner(value, name + key + '_')
+            elif type(_json) is list:
+                for i in range(len(_json)):
+                    __inner(_json[i], name + str(i) + '_')
+            else:
+                out[name[:-1]] = _json
+        __inner(obj)
+        return out
```

Comment:
Why do we need to flatten?

Reply:
Just for an easier look ups. The json output is nested. so "main::temp" becomes "main_temp" in dictionary lookup.

Reply:
The YahooWeather-widget does this, too. But to be fair, there is no other layer for reading out the values, the flattened json is made available to the user directly.